### DISCUSSION TITLE
howto get legendary & which-key to run with lazy

### DISCUSSION THREAD

**Discussion by svenXY** [mrjones2014/legendary.nvim#423] [Q&A]
Hi, in https://github.com/mrjones2014/smart-splits.nvim/discussions/135#discussioncomment-7455630 you said:

> The which-key.nvim extension for legendary.nvim is dependent on load order if you want to auto-load items into both, unfortunately there isn't really any way around that. If you open a discussion on the legendary.nvim repo, we can go from there, but the tl;dr: is the order of operations needed to make the which-key.nvim extension work properly is:
> 
> 1.     which-key.nvim loaded by plugin manager onto Neovim's runtimepath, but do not load the keymaps yet
> 2.     legendary.nvim loaded and initialized via require('legendary').setup(opts)
> 3.     Register your keymaps with which-key.nvim, now that both plugins are ready, which-key.nvim items will be auto-loaded into legendary.nvim

so here it is ;-)

I have no idea on how to load which-key in lazy without the keymaps nor how to then load them in the end, I'm afraid

**Comment by mrjones2014** [comment]
So, if you're using `lazy.nvim`, you should be able to achieve that by something like the following:

```lua
{
  'mrjones2014/legendary.nvim',
  dependencies = { 'folke/which-key.nvim' },
  config = function()
    -- load which-key and setup config
    local wk = require('which-key')
    wk.setup(opts)
    -- load legendary and register which-key extension
    local legendary = require('legendary')
    legendary.setup({
      -- other opts here
      extensions = {
        which_key = { auto_register = true },
      }
    })
    -- register keymaps with which-key, now they will
    -- be auto-loaded into legendary as well
    wk.register(your_keymap_tables)
  end
}
```

**Reply by svenXY** [reply]
wk.setup(opts) <- undefined global 'opts'

so, I have two files loaded by lazy.nvim, one is the legendary.lua we have here, the other is (mappings shortened for brevity)

```lua
return {
  "folke/which-key.nvim",
  event = "VeryLazy",
  config = function()
    vim.o.timeout = true
    vim.o.timeoutlen = 300
    local status_ok, wk = pcall(require, "which-key")
    if not status_ok then
      return
    end

    local setup = {
      icons = {
        breadcrumb = "»", -- symbol used in the command line area that shows your active key combo
        separator = "➜", -- symbol used between a key and it's label
        group = "+", -- symbol prepended to a group
      },
      ignore_missing = true, -- enable this to hide mappings for which you didn't specify a label
    }

    local opts = {
      mode = "n", -- NORMAL mode
      prefix = "<leader>",
      buffer = nil, -- Global mappings. Specify a buffer number for buffer local mappings
      silent = true, -- use `silent` when creating keymaps
      noremap = true, -- use `noremap` when creating keymaps
      nowait = true, -- use `nowait` when creating keymaps
    }

    local mappings = {
      ["a"] = { "<cmd>Alpha<cr>", "Alpha" },
      --[[ ["e"] = { "<cmd>NvimTreeToggle<cr>", "Explorer" }, ]]
      ["W"] = { "<cmd>w!<CR>", "Save Buffer" },
      --[[ ["q"] = { "<cmd>q!<CR>", "Quit Neovim" }, ]]
      ["c"] = { "<cmd>Bdelete!<CR>", "Close Buffer" },
      ["h"] = { "<cmd>nohlsearch<CR>", "No Highlight" },
      ["P"] = { "<cmd>Telescope project<CR>", "Projects" },

      f = {
        name = "Telescope",
        b = {
          "<cmd>Telescope buffers<cr>",
          "Buffers",
        },
        f = {
          "<cmd>Telescope find_files<cr>",
          "Find files",
        },
        h = { "<cmd>Telescope help_tags<cr>", "Help"},
        -- d = {'<cmd>lua require("user.telescope").search_dotfiles({ hidden = true })<CR>', "dotfiles"},
        -- D = {'<cmd>lua require("user.telescope").grep_dotfiles({ hidden = true })<CR>', "grep dot"},
        s = {
          "<cmd>Telescope luasnip theme=dropdown layout_config={width=0.8}<cr>",
          "Snippets",
        },
        t = { "<cmd>Telescope live_grep<cr>", "Find Text" },
        p = { "<cmd>Telescope neoclip<cr>", "Clipboard" },
        w = {'<cmd>lua require("plugins.telescope").grep_wiki({ hidden = true })<CR>', "wiki"},
        c = {'<cmd>lua require("user.telescope_extra").grep_buffer({ hidden = true })<CR>', "current buffer"},
      },
    }

    wk.setup(setup)
    wk.register(mappings, opts)
  end
}

```

**Reply by mrjones2014** [reply]
> wk.setup(opts) <- undefined global 'opts'

I didn't mean literally the variable `opts`, I just meant your options you pass to which-key 😆 

Basically, your `legendary.nvim` setup needs to happen in between your `wk.setup(setup)` and `wk.register(mappings, opts)` lines in your config there. You can do that by restructuring the way I suggested, or by just adding `require('legendary')` in between those lines, which should cause `lazy.nvim` to load `legendary` and run it's config.

**Reply by svenXY** [reply]
nope, still no success, I'm afraid.

I removed the wk.setup(opts) from legendary.lua and added a `require('legendary')` inbetween the last two lines of which-key.lua.

**Reply by mrjones2014** [reply]
Can you try restructuring it like I showed in the top comment?

**Reply by svenXY** [reply]
I can try.

Can I somehow load these opts and your_keymap_tables from the which-keys.lua file? I'd really like to keep them there.

Sorry, I'm not a lua developer...

**Reply by mrjones2014** [reply]
You could put them in a separate Lua module. Anything under `~/.config/nvim/lua` will be available on `runtimepath`, so if you put them at, for example, `~/.config/nvim/lua/keymaps.lua` then you would be able to `require('keymaps')`

**Reply by svenXY** [reply]
All right, this is much too complicated just for getting smart-splits to work - which was my initial problem. I have since solved this with normal keybindings.

As much as I like the idea of legendary, it is of not too much use for me without proper which-key integration, but the setup required here to get this to work is a bit too complicated for me, sorry. I was not initially looking for the functionality legendary provides and will most probably not miss not having it (especially since there is telescope support for mappings).

Thanks a lot for your support anyway.
